### ISSUE TITLE
DI bindings requires services to be public

### ISSUE THREAD

**Issue by ro0NL** [symfony/symfony#24559]
| Q                | A
| ---------------- | -----
| Bug report?      | yes?
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.4

Given

```yaml
foo:
    class: App\Foo
    public: true
    bind: { $some: '@bar' }
bar:
    class: stdClass
```

I get

```
The service "foo" has a dependency on a non-existent service "bar". 
```

Making `bar` public fixes it. Am i missing something why "bar" cant be private?

**Comment by weaverryan** [comment]
Yea, this seems wrong... I assume if you used arguments instead of bind, it works fine (i.e. the issue is special to bind)?

**Comment by chalasr** [comment]
I can confirm, it is special to bind.
Referencing the `bar` private service from another public service fixes it also:

```yaml
foo:
    class: App\Foo
    public: true
    bind: { $some: '@bar' }
bar:
    class: stdClass
public_1:
    public: true
    class: stdClass
    arguments: ['@bar']
```

I did not have a deeper look yet.

**Comment by dmaicher** [comment]
I tried to debug it a bit.

The problem happens when the definition of service `bar` is inlined into `foo`. In that case the argument 0 for `foo` is no longer a `Reference` but a `Definition`.

BUT the bindings of `foo` still contain a `Reference` to service `bar` which has been removed by `RemoveUnusedDefinitionsPass` as it was inlined.

Within `CheckExceptionOnInvalidReferenceBehaviorPass` all bindings of `foo` are then checked and the Reference to `bar` turns out to be invalid as it indeed does not exist on the container anymore.

Not sure though what's the best way to fix this :see_no_evil:
